dashboard/app/handler.go

```diff
@@ -339,3 +354,53 @@ func encodeCookie(w http.ResponseWriter, cd *cookieData) {
 }
 
 var templates = html.CreateGlob("*.html")
+
+type gzipResponseWriterCloser struct {
+	w   *gzip.Writer
+	buf *bytes.Buffer
+	rw  http.ResponseWriter
+}
+
+func (g *gzipResponseWriterCloser) Write(p []byte) (n int, err error) {
+	return g.w.Write(p)
+}
+
+func (g *gzipResponseWriterCloser) Close() {
+	if g.w != nil {
+		g.w.Close()
+	}
+}
+
+func (g *gzipResponseWriterCloser) Header() http.Header {
+	return g.rw.Header()
+}
+
+func (g *gzipResponseWriterCloser) WriteHeader(statusCode int) {
+	g.rw.WriteHeader(statusCode)
+}
+
+func (g *gzipResponseWriterCloser) writeResult(w http.ResponseWriter, writeUngzipped bool) (int, error) {
+	g.w.Close()
+	g.w = nil
+	if !writeUngzipped {
+		w.Header().Set("Content-Encoding", "gzip")
+		return w.Write(g.buf.Bytes())
+	}
+	r, err := gzip.NewReader(g.buf)
```

Comment:
You first compress the data, then uncompress. Does it have to be that complicated?

It must be possible to evaluate `!strings.Contains(r.Header.Get("Accept-Encoding"), "gzip")` before the first Write, so we can just not use `gzip.NewWriter` if it's not needed.

Reply:
I'd like to be able to return error codes in the middle of the content generation.
Writing directly to http.ResponseWriter we lose this opportunity.

Reply:
You are anyway writing to `g.buf` in this type, I think it's just the question of whether to use `w: addEmptyCloser{buf}` or `w: gzip.NewWriter(buf)`.

Reply:
Please add a comment that explains this.

Reply:
It is explained in https://github.com/google/syzkaller/issues/5731#issuecomment-2633451804.
I'll copy-paste it to the code.